File: 0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* prev = NULL;
    int count = 1;
    int max = -1;
    void getNodes(TreeNode* node, vector<int>& ans) {
        if(!node) return;
        
        getNodes(node->left,ans);
        if(prev) {
            if(node->val == prev->val) {
                count++;
            }
            else {
                count = 1;
            }
        }
        
        if(count>max) {
            max = count;
            ans.clear();
            ans.push_back(node->val);
        }
        else if(count == max) {
            ans.push_back(node->val);
        }
        
        prev = node;
        getNodes(node->right,ans);
    }
    vector<int> findMode(TreeNode* root) {
        if(!root) return {};
        vector<int> ans;
        getNodes(root,ans);
        return ans;
    }
};

File: 0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode prev = null;
    int count = 1;
    int max = -1;
    public void getNodes(TreeNode node,ArrayList<Integer> ans) {
        if(node==null) return;
        getNodes(node.left,ans);
        if(prev != null) {
            if(node.val == prev.val) {
                count++;
            }
            else {
                count = 1;
            }
        }
        
        if(count>max) {
            ans.clear();
            ans.add(node.val);
            max = count;
        }
        else if(count == max) {
            ans.add(node.val);
        }
        prev = node;
        getNodes(node.right,ans);
    }
    public int[] findMode(TreeNode root) {
        ArrayList<Integer> ans = new ArrayList<>();
        getNodes(root,ans);
        int[] retAns = new int[ans.size()];
        for(int i=0;i<retAns.length;i++){
            retAns[i] = ans.get(i);
        }
        return retAns;
    }
}